### ISSUE TITLE
Math is difficult to implement

### ISSUE THREAD

**Issue by timokoesters** [huggingface/candle#691]
In Python it's possible to write just this:
`3 * x**2 - 4*x + 5`

With Candle it currently looks like this, which is difficult to read and write:
`Ok((&(&(&x.powf(2.0)? * 3.0)? - &x * 4.0)? + 5.0)?)`

1. Why does everything return a Result? I assume the reason is that the GPU connection might fail, but in that case maybe it's okay to panic the whole application?
2. Why is Tensor not Copy? It's an Arc, so it would be okay in theory.
3. A few more traits can be implemented to make it look like this: `3.0 * x * x - 4.0 * x - 5.0`

I wrote an example here: 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e5abf73c9e38fb6cef2c17b5a1898a29

**Comment by LaurentMazare** [comment]
Yes the syntax is cumbersome, we have some ideas that should slightly improve it though it will not be as simple as the version you suggest. To answer your points:
1. All the ops may indeed fail, the most common thing is not really the GPU going down but more the GPU running out of memory, using invalid shape, etc.
2. I don't think a `Tensor` could be copy, copy implies that you can just do copy the immediate memory representation (see [here](https://doc.rust-lang.org/std/marker/trait.Copy.html)) which isn't the case for `Arc`. Hence `Tensor` is only clone - and the clone op is pretty inexpensive.

**Comment by LaurentMazare** [comment]
Both of these are actually already done :)

**Comment by timokoesters** [comment]
Oh that's great, maybe I was using an old version, or it only worked one way.

**Comment by timokoesters** [comment]
It looks like Tensor + Result is implemented, but not Result + Tensor or Result + Result.

I have another related problem: I'm unable to write a rusty variant for modifying a single index:
var[2, 3] += 1, is there a method that can help me with that?

**Comment by LaurentMazare** [comment]
Re `Result + Result` I think this is because of the foreign trait limitation of rust so not much that we can do here.
Re assigning single values in place, we don't really have an op to do this and it's on purpose as these tend to be really inefficient so we prefer favoring more functional approaches and avoiding accessing data by index. I don't think this has been an issue for all the models we've been looking at so far but if there are some specific models for which this would be a limitation, certainly curious to hear about them.

**Comment by timokoesters** [comment]
Sorry, you are correct on the Result+Result, but the other combinations will work, as you can see in my example. An idea to even allow result+result is to have all operations return Tensor, but hide the Result internally and only expose it with some getter methods.

I am redoing the "Neural networks from zero to hero" with Candle and one of the first things is building a character combination count matrix (https://youtu.be/PaCmpygFfXo?feature=shared&t=574).

Anyway, thanks for your quick answers!

**Comment by timokoesters** [comment]
> most common thing is not really the GPU going down but more the GPU running out of memory, using invalid shape

These sound like unrecoverable errors, so it could be okay to panic in these cases?

**Comment by chronicl** [comment]
Just to chime in some ideas/opinions:
To somewhat alleviate the trouble with rusts foreign trait limitation it's possible to create create a wrapper type that can be used to wrap any type we want to be able to use as the lhs of an operation, for example like this
```rust
use candle::Result;

struct Wrapper<T>(T);

trait Wrap: Sized {
    fn w(self) -> Wrapper<Self> {
        Wrapper(self)
    }
}

impl Wrap for f64 {}

impl<T> Add<Tensor> for Wrapper<T>
where
    Tensor: Add<T, Output = Result<Tensor>>,
{
    type Output = Result<Tensor>;
    fn add(self, rhs: Tensor) -> Self::Output {
        rhs + self.0
    }
}

let t = Tensor::new(&[2f32], &Device::Cpu).unwrap();
let result = 1.0.w() + t;
```
this still isn't very ergonomic because after the first operation you get a `Result` so
```rust
let result = 2.0.a() + t + 3.0;
```
doesn't work. Instead you would need to do
```rust
(2.0.a() + t)? + 3.0;
```
or you also implement `Wrap` for `Result<Tensor>`, which doesn't gain us much though:
```rust
(2.0.a() + t).a() + 3.0;
```
I do like this idea:
> hide the Result internally and only expose it with some getter methods.

this would allow for more ergonomic math and allow users to choose when they want to return early due to an error (if at all). I believe Rust's [track_caller](https://rustc-dev-guide.rust-lang.org/backend/implicit-caller-location.html#codegen-examples) can be used to track the location of where the error actually happened, so we wouldn't lose that information 

I haven't felt the need for tensors being `Copy` personally, but it could be achieved by having some global state hold all the information of the tensors and making the tensors themselves ids for retrieving that information. This is commonly done in rust whenever you want a type to be `Copy` for ergonomic reasons, for example in [sycamore's](https://github.com/sycamore-rs/sycamore) reactive systems.

**Comment by chronicl** [comment]
Making tensors `Copy` could be achieved by having some global state hold all the information of the tensors and making the tensors themselves ids for retrieving that information. This is commonly done in rust whenever you want a type to be `Copy` for ergonomic reasons, for example in [sycamore's](https://github.com/sycamore-rs/sycamore) reactive systems. Of course this comes along with all of the caveats of dealing with global state.

I do like this idea:
> hide the Result internally and only expose it with some getter methods.

this would allow for more ergonomic math and allow users to choose when they want to return early due to an error (if at all). I believe Rust's [track_caller](https://rustc-dev-guide.rust-lang.org/backend/implicit-caller-location.html#codegen-examples) can be used to track the location of where the error actually happened, so we wouldn't lose that information.

**Comment by Narsil** [comment]
100% on this. Having implemented a few servers, panicking is not great !

**Comment by cchance27** [comment]
Trying my first hand at math with candle, and just ran into something... is Add<Tensor> only available really for f64? Not u32, etc?

**Comment by LaurentMazare** [comment]
> Trying my first hand at math with candle, and just ran into something... is Add only available really for f64? Not u32, etc?

Right, it's just a convenient shortcut. If you want to add an u32, you can do something like `tensor.broadcast_add(&Tensor::new(42u32, tensor.device())?)?`.

**Comment by cchance27** [comment]
Not gonna lie, I’m sorta an idiot as I figured I could use broadcast_add but then was like “wtf it takes a tensor, not a u32…. I’m so confused… completely forgetting I could just create a single value tensor lol

**Comment by asukaminato0721** [comment]
Oh, I currently use `tensor.affine(1., 42.)?` lol.

**Comment by timokoesters** [comment]
I've worked on a wrapper type around Result<Tensor> called MTensor ("Maybe Tensor") that allows simpler math: https://github.com/timokoesters/candle/blob/f457844f6f0435f748a2967b9b04d0f554adfd37/candle-core/src/mtensor.rs#L211-L222

@LaurentMazare Can you take a look? Is that something that could be merged into Candle?

PS: It relies on this unstable feature https://github.com/rust-lang/rust/issues/84277, hopefully it will work in stable rust soon.

**Comment by ivarflakstad** [comment]
Hey @timokoesters !

I don't think this is a bad idea at all, but I don't expect that rust feature to be stabilised any time soon.

I think we can improve the API with just a newtype + asref/deref to result though.

It would enable us to implement the missing `MTensor + Tensor` and `MTensor + MTensor`, but we would have to set `type Output = MTensor` in the `Add`.
For each op we would check the value of `MTensor`. If it's an `Err` then we just assign that as the value (not sure what to so if both are `Err` hehe) and move on.
If it's `Ok` then we run the op.
When we want to finally extract our `Tensor` we use `let value = mtensor.into()?` or whatever.

I think I prefer a name like `TensorResult` over `MTensor`, personally.

**Comment by ivarflakstad** [comment]
> not sure what to so if both are Err 

Nvm we'd just aggregate them. Easy.
